### PR TITLE
discover CR policy and binding  for container vertical scale in kubeturbo

### PR DESCRIPTION
**Pull request by libai98**
1) implement code to discovery CR policy and binding  for container vertical scale in kubeturbo
2) add code to convert discovered settings into Turbo policy

### CODE REVIEW

File: pkg/discovery/processor/turbo_policy_processor.go [hunk 1]
```diff
@@ -26,13 +26,20 @@ func (p *TurboPolicyProcessor) ProcessTurboPolicies() {
 	turboSloScalings, err := p.ClusterScraper.GetAllTurboSLOScalings()
 	if err != nil {
 		glog.Warningf("Failed to list SLOHorizontalScales: %v.", err)
-		return
 	}
-	if len(turboSloScalings) == 0 {
-		glog.V(2).Info("There is no SLOHorizontalScale resource found in the cluster.")
+
```
Comment (pankajbatra1) [inline pkg/discovery/processor/turbo_policy_processor.go:30]: Please generalize and put logging in a separate function.  This should not need to change if new policies are supported in future.

File: pkg/discovery/processor/turbo_policy_processor.go [hunk 2]
```diff
@@ -46,17 +53,23 @@ func (p *TurboPolicyProcessor) ProcessTurboPolicies() {
...
 
 	policyMap := make(map[string]*repository.TurboPolicy)
+
 	for _, sloScale := range turboSloScalings {
-		// Create a copy as sloScale variable is reused during range loop
-		sloScaleCopy := sloScale
-		gvk := sloScaleCopy.GetObjectKind().GroupVersionKind()
-		if gvk.Empty() {
-			continue
+		if gvk := sloScale.GetObjectKind().GroupVersionKind(); !gvk.Empty() {
```
Comment (pankajbatra1) [inline pkg/discovery/processor/turbo_policy_processor.go:58]: Please see if you can refactor the code inside the two loops here

File: pkg/discovery/worker/group_discovery_policy_cvs.go [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+		if !validWorkloadController(target.Kind) {
+			return nil, fmt.Errorf("target %v is not a workload controller", target)
+		}
+
+		controllerRegex := target.Name
+		containerRegex := target.Container
+		namespace := policyBinding.GetNamespace()
+
+		for key, value := range worker.cluster.ControllerMap {
+			fmt.Println(key, value)
```
Comment (pankajbatra1) [inline pkg/discovery/worker/group_discovery_policy_cvs.go:59]: Is this needed?

File: pkg/discovery/worker/group_discovery_policy_cvs.go [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+	"github.com/turbonomic/turbo-go-sdk/pkg/proto"
+)
+
+func buildContainerVerticalScalePolicy(
+	worker *k8sEntityGroupDiscoveryWorker,
+	policyBinding *repository.TurboPolicyBinding) (*proto.GroupDTO, error) {
+	// Resolve group members
+
+	resolvedIds, err := resolveCVSPolicyTargets(worker, policyBinding)
+	if err != nil {
```
Comment (pankajbatra1) [inline pkg/discovery/worker/group_discovery_policy_cvs.go:19]: Do you want to log something about resolved IDs for a policy?

File: pkg/discovery/worker/group_discovery_worker.go [hunk 5]
```diff
@@ -255,155 +247,10 @@ func (worker *k8sEntityGroupDiscoveryWorker) BuildTurboPolicyDTOsFromPolicyBindi
...
-// validateActionMode validates the action mode
-// Valid action mode must exist in actionModes map
-func validateActionMode(actionMode v1alpha1.ActionMode) (string, error) {
-	if mode, ok := actionModes[actionMode]; ok {
-		return mode, nil
-	}
-	return "", fmt.Errorf("unsupported action mode %v", actionMode)
+func (worker *k8sEntityGroupDiscoveryWorker) buildContainerVerticalScalePolicy(
+	policyBinding *repository.TurboPolicyBinding) (*proto.GroupDTO, error) {
+	return buildContainerVerticalScalePolicy(worker, policyBinding)
```
Comment (pankajbatra1) [inline pkg/discovery/worker/group_discovery_worker.go:255]: How is 'err' from buildContainerVerticalScalePolicy sent up to caller

File: pkg/discovery/worker/group_discovery_worker.go [hunk 6]
```diff
@@ -239,7 +224,14 @@ func (worker *k8sEntityGroupDiscoveryWorker) BuildTurboPolicyDTOsFromPolicyBindi
...
 			groupDTO, err := worker.buildSLOHorizontalScalePolicy(policyBinding)
 			if err != nil {
-				glog.Errorf("Failed to build %v policy: %v.", policyBinding.GetPolicyType(), err)
+				glog.Warningf("Failed to build SLO horizontal scale policy: %v", err)
+				continue
+			}
+			groupDTOs = append(groupDTOs, groupDTO)
+		case ContainerVerticalScale:
+			groupDTO, err := worker.buildContainerVerticalScalePolicy(policyBinding)
+			if err != nil {
```
Comment (pankajbatra1) [inline pkg/discovery/worker/group_discovery_worker.go:233]: Error is lost here as well.  Maybe this is something for the future.

Feel free to drop if this is likely to be picked up with status work.

File: pkg/discovery/processor/turbo_policy_processor.go (hunk 1 above)
Comment (libai98) [reply pkg/discovery/processor/turbo_policy_processor.go:30]: will address this in https://vmturbo.atlassian.net/browse/OM-98560.

File: pkg/discovery/processor/turbo_policy_processor.go (hunk 2 above)
Comment (libai98) [reply pkg/discovery/processor/turbo_policy_processor.go:58]: will try to address this in https://vmturbo.atlassian.net/browse/OM-98560.

File: pkg/discovery/worker/group_discovery_policy_cvs.go (hunk 4 above)
Comment (libai98) [reply pkg/discovery/worker/group_discovery_policy_cvs.go:21]: will try to address this in https://vmturbo.atlassian.net/browse/OM-98560

File: pkg/discovery/worker/group_discovery_worker.go (hunk 5 above)
Comment (libai98) [reply pkg/discovery/worker/group_discovery_worker.go:255]: will try to address this in https://vmturbo.atlassian.net/browse/OM-98560

File: pkg/discovery/worker/group_discovery_policy_cvs.go (hunk 3 above)
Comment (libai98) [reply pkg/discovery/worker/group_discovery_policy_cvs.go:61]: Removed. Will try to address adding more log in https://vmturbo.atlassian.net/browse/OM-98560

**Pull request merged**